### PR TITLE
Add wire tap pattern sample

### LINKED ISSUE #9: Create a sample for Wire Tap pattern
Create a sample for [Wire Tap](https://www.enterpriseintegrationpatterns.com/patterns/messaging/WireTap.html) pattern.

### PR DESCRIPTION
**Pull request by ushirask**
Fixes #9

### CODE REVIEW

File: wire-tap/wire-tap.bal [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+type WireTapRequest record {|
+    string 'table;
+    "INFO"|"WARNING"|"ERROR" severity;
+    string message;
+|};
+
+type StockResponse record {
+    string parentHandlingUnitUUID;
+    string StockItemUUID;
```
Comment (LakshanWeerasinghe) [inline wire-tap/wire-tap.bal:13]: inconsistency with the naming conventions.

File: wire-tap/wire-tap.bal [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    string EWMWarehouse;
+    string HandlingUnitNumber;
+    string ShelfLifeExpirationDate;
+    string CountryOfOrigin;
+};
+
+http:Client wh = check new("http://api.sap.com.balmock.io");
```
Comment (LakshanWeerasinghe) [inline wire-tap/wire-tap.bal]: use a name like `sapClient`.

File: wire-tap/wire-tap.bal [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+service /warehouse on new http:Listener(8080) {
+    
+    resource function get stock(string parentId, string productId) returns StockResponse|error {
+        StockResponse result = check wh->/WarehouseAvailableStock/[parentId]/[productId];
```
Comment (LakshanWeerasinghe) [inline wire-tap/wire-tap.bal]: Use hyphens is better.
Also the calling resource should be in plural form.
```suggestion
        StockResponse result = check wh->/warehouse\-available\-stocks/[parentId]/[productId];
```

File: wire-tap/wire-tap.bal (hunk 1 above)
Comment (ushirask) [reply wire-tap/wire-tap.bal:13]: Will have to use the names used in the json response here

File: wire-tap/wire-tap.bal (hunk 3 above)
Comment (ushirask) [reply wire-tap/wire-tap.bal]: This is the path used with the endpoint, so we should stick to that right?

File: wire-tap/wire-tap.bal [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    string role = "logger";
+};
+
+type WireTapRequest record {|
+    string 'table;
+    "INFO"|"WARNING"|"ERROR" severity;
+    string message;
+|};
+
+type StockResponse record {
```
Comment (LakshanWeerasinghe) [inline wire-tap/wire-tap.bal:11]: better to use closed records.

File: wire-tap/wire-tap.bal [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        StockResponse result = check wh->/WarehouseAvailableStock/[parentId]/[productId];
+        worker w returns error? {
+            check wiretap({'table: "stock", severity: "INFO", message: result.toString()});
+        }
+        return result;
+    }
+}
+
+
+function wiretap (WireTapRequest wt) returns error? {
```
Comment (LakshanWeerasinghe) [inline wire-tap/wire-tap.bal]:
```suggestion
function wiretap(WireTapRequest wt) returns error? {
```

File: wire-tap/wire-tap.bal (hunk 4 above)
Comment (ushirask) [reply wire-tap/wire-tap.bal:11]: Response may have other non-required fields

File: wire-tap/wire-tap.bal [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+type StockResponse record {
+    string ParentHandlingUnitUUID;
+    string StockItemUUID;
+    string EWMWarehouse;
+    string HandlingUnitNumber;
+    string ShelfLifeExpirationDate;
+    string CountryOfOrigin;
+};
+
+http:Client sapClient = check new("http://api.sap.com.balmock.io");
```
Comment (DimuthuMadushan) [inline wire-tap/wire-tap.bal]:
```suggestion
final http:Client sapClient = check new("http://api.sap.com.balmock.io");
```

File: wire-tap/wire-tap.bal [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        return result;
+    }
+}
+
+
+function wiretap(WireTapRequest wt) returns error? {
+    http:Client db = check new("http://api.snowflake.com.balmock.io");
+    SnowflakeRequest snowflakeRequest = {statement: string `insert into ${wt.'table} values (${wt.message}, ${wt.severity}))`};
+    json _ = check db->/statements.post(snowflakeRequest);
```
Comment (DimuthuMadushan) [inline wire-tap/wire-tap.bal:39]:
```suggestion
    _ = check db->/statements.post(snowflakeRequest);
```

Comment (ushirask) [reply wire-tap/wire-tap.bal:39]: There's a compiler error without `json`

File: wire-tap/wire-tap.bal [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        StockResponse result = check sapClient->/WarehousePhysicalStockProducts/[parentId]/[productId];
+        worker w returns error? {
+            check wiretap({'table: "stock", severity: "INFO", message: result.toString()});
+        }
+        return result;
+    }
+}
+
+
+function wiretap(WireTapRequest wt) returns error? {
```
Comment (manuranga) [inline wire-tap/wire-tap.bal]: `wiretap(string 'table", "INFO"|"WARNING"|"ERROR" severity, string message)`

**Comment by ushirask** [comment]
WarehousePhysicalStockProducts API reference https://api.sap.com/api/CE_WHSEPHYSICALSTOCKPRODUCTS_0001/resource/Warehouse_Physical_Stock_by_Product

File: wire-tap/wire-tap.bal [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        StockResponse result = check sapClient->/WarehousePhysicalStockProducts/[parentId]/[productId];
+        worker w returns error? {
+            check wiretap("stock", "INFO", result.toString());
+        }
+        return result;
+    }
+}
+
+
+function wiretap(string 'table, "INFO"|"WARNING"|"ERROR" severity, string message) returns error? {
```
Comment (manuranga) [inline wire-tap/wire-tap.bal]: Let's use `tableName` get get rid of `'`.
Let's add a type `LogLevel`.

**Review by manuranga — APPROVED** [review]

**Pull request merged**